openhcl/virt_mshv_vtl/src/processor/hardware_cvm/mod.rs

```diff
@@ -900,6 +907,102 @@ impl<T, B: HardwareIsolatedBacking> hv1_hypercall::ModifyVtlProtectionMask
     }
 }
 
+impl<T, B: HardwareIsolatedBacking> hv1_hypercall::TranslateVirtualAddressX64
+    for UhHypercallHandler<'_, '_, T, B>
+{
+    fn translate_virtual_address(
+        &mut self,
+        partition_id: u64,
+        vp_index: u32,
+        control_flags: hvdef::hypercall::TranslateGvaControlFlagsX64,
+        gva_page: u64,
+    ) -> HvResult<hvdef::hypercall::TranslateVirtualAddressOutput> {
+        if partition_id != hvdef::HV_PARTITION_ID_SELF {
+            return Err(HvError::AccessDenied);
+        }
+
+        if vp_index != hvdef::HV_VP_INDEX_SELF && vp_index != self.vp.vp_index().index() {
+            return Err(HvError::AccessDenied);
+        }
+
+        let target_vtl = self
+            .target_vtl_no_higher(
+                control_flags
+                    .input_vtl()
+                    .target_vtl()?
+                    .unwrap_or(self.intercepted_vtl.into()),
+            )
+            .map_err(|_| HvError::AccessDenied)?;
+
+        if self.intercepted_vtl == target_vtl {
+            return Err(HvError::AccessDenied);
+        }
+
+        let gva = gva_page * hvdef::HV_PAGE_SIZE;
+
+        let registers = self.vp.backing.translation_registers(self.vp, target_vtl);
+
+        let pat = self
+            .vp
+            .access_state(target_vtl.into())
+            .cache_control()
```

Comment:
This is really expensive. Can we just get the PAT without getting the rest of the MTRRs?